### ISSUE TITLE
npm install --production modifies package-lock.json

### ISSUE THREAD

**Issue by jyotman** [npm/npm#16967]
#### I'm opening this issue because:

  - [ ] npm is crashing.
  - [ ] npm is producing an incorrect install.
  - [X] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?
`npm install --production` modifies package-lock.json. If I do `npm install` without the production flag then it installs all the packages and package-lock.json is not modified. Basically it removes all entries from the package-lock.json file which have the `dev: true` attribute present.
#### How can the CLI team reproduce the problem?
1. Create a package.json file which has both dependencies and devDependencies. 
2. Generate a package-lock.json file by running `npm install`.
3. Delete the node-modules folder.
4. Now run `npm install --production`. This step will modify the package-lock.json.

### supporting information:

 - `npm -v` prints: 5.0.2
 - `node -v` prints: 7.6.0
 - `npm config get registry` prints: https://registry.npmjs.org/
 - Windows, OS X/macOS, or Linux?: Linux

**Comment by boneskull** [comment]
I think this has caused CI failures in Mocha.  First, we run a sanity check to ensure the package can be installed by a user:

```bash
$ npm install --production
# package-lock.json is created at this point
$ npm run sanity-check
```

(we have `package-lock=false` set in `.npmrc`)

Then, we install our dev deps and run the full suite:

```bash
$ npm install
# whoops, package-lock.json is preventing this from happening
$ npm test
# fails b/c lack of deps
```

The (untested) workaround is to `rm -f package-lock.json`.

**Comment by boneskull** [comment]
After trying my workaround, this is not the issue, though I don't know what is.  I am missing a dependency.  [Here's](https://travis-ci.org/mochajs/mocha/jobs/240501885) a broken build job, if someone is interested.

**Comment by legodude17** [comment]
Does the same issue occur if you run `npm i --production --no-package-lock`?

**Comment by boneskull** [comment]
My latest theory:

I think it may have been a mismatch between what I thought the package name was and what the actual package name was.  For example:

```json
"dependencies": {
  "foo": "https://github.com/somebody/foo.git"
}
```

If that is a valid package URL, but the `package.json` in that repo has the name `@somebody/foo` instead of the expected 'foo', then npm@5 seems to have some problems that npm@4 did not.

**Comment by boneskull** [comment]
The above is *not* the same issue reported by the OP, though I can reproduce it.  I'll create a new issue for this.

**Comment by zkat** [comment]
This is fixed by #17508. It's available in the npm canary (`npm i -g npmc@latest`), and will be included in the next npm release next week.
